dialogy/workflow/workflow.py

```diff
@@ -258,62 +200,30 @@ def execute(self) -> Workflow:
                 history = {
                     "plugin": plugin,
                     "before": {
-                        "input": self.input,
-                        "output": self.output,
+                        "input": input.dict(),
+                        "output": output.dict(),
                     },
                 }
+
             start = time.perf_counter()
-            plugin(self)
+            with self.lock:
+                input, output = plugin(input, output)
             end = time.perf_counter()
+
             # logs are available only when debug=False during class initialization
             if self.debug:
                 history["after"] = {
-                    "input": self.input,
-                    "output": self.output,
+                    "input": input.dict(),
+                    "output": output.dict(),
                 }
                 history["perf"] = round(end - start, 4)
-            if history:
-                logger.debug(history)
-        return self
-
-    def run(self, input_: Input) -> Tuple[Dict[str, Any], Dict[str, Any]]:
-        """
-        .. _workflow_run:
-
-        Get final results from the workflow.
 
-        The current workflow exhibits the following simple procedure:
-        pre-processing -> inference -> post-processing.
-        """
-        with self.lock:
-            self.input = input_
-            try:
-                return self.execute().flush()
-            finally:
-                self.__reset()
-
-    def flush(self) -> Tuple[Dict[str, Any], Dict[str, Any]]:
-        """
-        Reset :code:`workflow.input` and :code:`workflow.output`.
-        """
-        if self.input is None or self.output is None:
-            return {}, {}
-        input_ = copy.deepcopy(self.input.json())
-        output = copy.deepcopy(self.output.json())
-        self.__reset()
-        return input_, output
-
-    def json(self) -> Dict[str, Any]:
-        """
-        Represent the workflow as a python dict.
+            if history:
+                logger.debug(pformat(history))
+        pipeline_end = time.perf_counter()
+        output.duration = round(pipeline_end - pipeline_start, 3)
 
-        :rtype: Dict[str, Any]
-        """
-        return attr.asdict(
-            self,
-            filter=lambda attribute, _: attribute.name
-            not in self.NON_SERIALIZABLE_FIELDS,
-        )
```

Comment:
This should stay right? We aren't using pydantic's `BaseModel` here.

On second thought, we should just get rid of `attr` here as well, right?

Reply:
Can get rid of attr here since we don't need `Workflow` to be a data class.